Add configuration parameter to specify a path to a custom postgresql.base.conf and disable its backup

zenitraM opened the pull request:
This adds a `custom_conf` parameter on the `postgresql` settings to specify a path for a custom configuration (outside the data path), that replaces postgresql.base.conf and is `include`d on its place on the real postgresql.conf.

Setting this parameter also disables the handling and backup of postgresql.base.conf - it is expected that the user will be responsible of ensuring the external configuration file exists if this parameter is set.

It will also disable pg_hba backup if a custom `hba_file` location is set on `parameters`.

Related to #81, #53.

CyberDem0n commented on patroni/postgresql.py:
```diff
@@ -100,9 +100,13 @@ def __init__(self, config):
-        self._configuration_to_save = (self._postgresql_conf, self._postgresql_base_conf,
-                                       os.path.join(self._data_dir, 'pg_hba.conf'))
+        self._configuration_to_save = [self._postgresql_conf]
+        if not self._postgresql_custom_conf:
```
`postgresql.custom_conf` and `postgresql.parameters.hba_conf` can be set, updated and removed from the configuration in a run-time either by updating of `patroni.yaml` config file manually and sending SIGHUP or by changing value of the cluster `config` in DCS.

`_configuration_to_save` if used to make a copies of the configuration files to be backed up by wal-e (it skips all *.conf files)
In case of external configuration files I think it is still make sense to back them up by copying `custom_conf` and `hba_file` into `data_dir` with the names "postgresql.base.conf.backup" and "pg_hba,conf.backup" accordingly.
 
Such functionality requires some changes in a `save_configuration_files` and `restore_configuration_files` methods and we will need to change the way we keep list of configuration files from `list` or `tuple` to `dict`.

`Postgresql.reload_config` method should update value `_configuration_to_save` if the `custom_conf` or `hba_file` were changed.

CyberDem0n replied:
> Also, when being restored on another instance that also has custom_conf and hba_file, what do we want to do? overwrite the file at the custom_conf path with the conf that was carried over, or copy it as postgresql.base.conf but keep it unused?.

The second option seems reasonable to be, but taking into account following:
> I find it cleaner to just explain: if you use custom_conf or hba_file, they will be under your control, there will not be any postgresql.base.conf and the configuration files will not be carried over on base_backups, and it's your responsability to ensure the settings on them are consistent and the files exist on all nodes.

I think you are right. We need document this behavior and let user live with it. Actually this is not the only drawback. Also I would like to add: currently Patroni is  "tracking" dynamic changes of postgres parameters and tells you in the API that restart is needed to apply some of the changes. In case of external configuration file it wont tell you this information..

CyberDem0n commented on patroni/postgresql.py:
```diff
@@ -622,12 +626,12 @@ def restart(self):
     def _write_postgresql_conf(self):
         # rename the original configuration if it is necessary
-        if not os.path.exists(self._postgresql_base_conf):
+        if not self._postgresql_custom_conf and not os.path.exists(self._postgresql_base_conf):
```
What if there is no `postgresql.conf` in the data directory? I think it is exactly your case...

zenitraM replied:
This prevents a `postgresql.base.conf` to be created from an original `postgresql.conf` if there is a custom conf somewhere else, as it will not be used.

The following lines will overwrite the `postgresql.conf` on the data dir anyway.

CyberDem0n replied:
ok

zenitraM commented:
I've moved `_configuration_to_save` to its own property method (so I think there's no need to handle anything special on reload now).
 I've also added some mentions to the drawbacks of using its own base, but not sure where in the docs would the behavior change about the `hba_file` parameter fit.

CyberDem0n commented on patroni/postgresql.py:
```diff
@@ -100,9 +100,8 @@ def __init__(self, config):
         self._postgresql_conf = os.path.join(self._data_dir, config_base_name + '.conf')
         self._postgresql_base_conf_name = config_base_name + '.base.conf'
         self._postgresql_base_conf = os.path.join(self._data_dir, self._postgresql_base_conf_name)
+        self._postgresql_custom_conf = config.get('custom_conf')
```
I think you need to convert it into 'property' as well, otherwise it can't be changed dynamically...

zenitraM replied:
:+1:, just changed it to always be queried dynamically

CyberDem0n commented on patroni/postgresql.py:
```diff
@@ -125,6 +123,15 @@ def __init__(self, config):
     @property
+    def _configuration_to_save(self):
+        configuration = [self._postgresql_conf]
+        if not config.get('custom_conf'):
```
if not self.config

CyberDem0n commented on patroni/postgresql.py:
```diff
@@ -622,12 +629,12 @@ def restart(self):
     def _write_postgresql_conf(self):
         # rename the original configuration if it is necessary
-        if not os.path.exists(self._postgresql_base_conf):
+        if not config.get('custom_conf') and not os.path.exists(self._postgresql_base_conf):
```
if 'custom_conf' not in self.config and not os.path.exists(self._postgresql_base_conf):

CyberDem0n commented on patroni/postgresql.py:
```diff
@@ -622,12 +629,12 @@ def restart(self):
         with open(self._postgresql_conf, 'w') as f:
             f.write('# Do not edit this file manually!\n# It will be overwritten by Patroni!\n')
-            f.write("include '{0}'\n\n".format(self._postgresql_base_conf_name))
+            f.write("include '{0}'\n\n".format(config.get('custom_conf') or self._postgresql_base_conf_name))
```
f.write("include '{0}'\n\n".format(self.config.get('custom_conf') or self._postgresql_base_conf_name))

CyberDem0n commented:
:+1:

alexeyklyukin commented:
👍
